src/coroutine.rs

```diff
@@ -0,0 +1,157 @@
+//! Python coroutine implementation, used notably when wrapping `async fn`
+//! with `#[pyfunction]`/`#[pymethods]`.
+use crate::coroutine::waker::AsyncioWaker;
+use crate::exceptions::{PyRuntimeError, PyStopIteration};
+use crate::pyclass::IterNextOutput;
+use crate::types::PyIterator;
+use crate::{IntoPy, Py, PyAny, PyErr, PyObject, PyResult, Python};
+use pyo3_macros::{pyclass, pymethods};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+mod cancel;
+mod waker;
+
+pub use crate::coroutine::cancel::{CancelHandle, CoroutineCancel};
+
+const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
+
+/// Python coroutine wrapping a [`Future`].
+#[pyclass(crate = "crate")]
+pub struct Coroutine {
+    future: Option<Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>>,
+    cancel: Option<CancelHandle>,
+    waker: Option<Arc<AsyncioWaker>>,
+}
+
+impl Coroutine {
+    ///  Wrap a future into a Python coroutine.
+    ///
+    /// Coroutine `send` polls the wrapped future, ignoring the value passed
+    /// (should always be `None` anyway).
+    ///
+    /// `Coroutine `throw` drop the wrapped future and reraise the exception passed
+    pub fn from_future<F, T, E>(future: F) -> Self
+    where
+        F: Future<Output = Result<T, E>> + Send + 'static,
+        T: IntoPy<PyObject> + Send,
+        E: Send,
```

Comment:
Where does the send bound on `E` come from?

Reply:
Nowhere, I will remove it (`Send`bound on `T` is also useless)